docs/sources/lids/0003-Coprocessor.md

```diff
@@ -0,0 +1,113 @@
+---
+title: "0003: Coprocessor"
+description: "introduce coprocessor from Google’s BigTable coprocessor and HBase coprocessor"
+---
+
+# 0003: Coprocessor
+
+**Author:** liguozhong (fuling.lgz@alibaba-inc.com)
+
+**Date:** 02/2023
+
+**Sponsor(s):** @jeschkies
+
+**Type:** Feature
+
+**Status:** Draft
+
+**Related issues/PRs:** https://github.com/grafana/loki/issues/8559 AND https://github.com/grafana/loki/issues/91
+
+**Thread from [mailing list](https://groups.google.com/forum/#!forum/lokiproject):** N/A
+
+---
+
+## Background
+
+{log_type="service_metrics"} |= "ee74f4ee-3059-4473-8ba6-94d8bfe03272"
+
+We have counted the source distribution of our logql, and 85% of the grafana log explore queries are traceID queries.
+Generally, the log time range of traceID is about 10 minutes（trace time= start～end）
+
+## Problem Statement
+
+but because users do not know traceID start time and end time , they usually search for 7 day log. 
+In fact having a time range of "7d-10m" is an invalid search.
+
+## Goals
```

Comment:
If I understood the HBade use case they basically created a plug-in system with some hooks. This seems to be the goal here. How would we implement it? How are we shipping the data? We could use Hashicorp's go-plug but there's probably s serialization overhead. As I understood the HBade design the coprocessor is started on the host where the data is an this can read the data directly. Not sure we can do it here.

Reply:
> If I understood the HBade use case they basically created a plug-in system with some hooks.

yes.

> How would we implement it? How are we shipping the data? We could use Hashicorp's go-plug but there's probably s serialization overhead

My suggestion is that we should not restrict the language. Through http+protobuf, more developer can work together with loki. 
Because prometheus remote read and k8s hpa implement the plugin mechanism through http+protobuf, this  has achieved great success.
The implementation of HBase limits the jvm language so that the coprocessor can only be implemented in languages such as java or scala. I don't think it's a good idea.



> Hashicorp's go-plug

Will this limit the development language?

Reply:
> Will this limit the development language?

No, they also use [gRPC](https://github.com/hashicorp/go-plugin#features) but make it a little simpler for plugins in Go.

Reply:
It would be great if there is no restriction on the implementation language. 
Can we refer to open telemetry’s collector that supports both http and grpc protocols?
The SREs I know actually don’t know much about grpc, but they can easily build an http server .

Reply:
I don't think the plug-in is restricted to Go. One could use gRPC.

Reply:
great